mitreattack/navlayers/generators/usage_generator.py

```diff
@@ -37,113 +45,55 @@ def __init__(self, source, matrix='enterprise', local=None):
+        raise UnableToFindStixObject
 
-    def get_mitigation(self, match):
+    def get_matrix_data(self, match_pattern):
         """
-        Retrieve mitigation for a given match
-        :param match: matching pattern
-        :return: first matching mitigation or UnableToFindMitigation exception
-        """
-        mitigation_list = self.source_handle.query([Filter('type', '=', 'course-of-action')])
-        for x in mitigation_list:
-            for y in x['external_references']:
-                if y['source_name'] == 'mitre-attack' and y['external_id'] == match:
-                    return x
-            if match in x['name']:
-                return x
-        raise UnableToFindMitigation
-
-    def get_matrix_data(self, match_pattern, obj_type):
-        """
-        Retrieve matching attack-pattern objects for match_pattern
+        Retrieve a list of matching attack-pattern (technique) objects for a match_pattern (group, software, mitigation)
         :param match_pattern: the pattern to match
-        :param obj_type: the type of object match_pattern is
         :return: list of associated attack-pattern objects
         """
         out = []
-        if obj_type == 'group':
-            gr = self.get_group(match_pattern)
-            related = self.source_handle.relationships(gr, 'uses', source_only=True)
+        obj = self.get_stix_object(match_pattern)
+        if obj.type == 'group':
+            related = self.source_handle.relationships(obj, 'uses', source_only=True)
             out = self.source_handle.query([Filter('type', '=', 'attack-pattern'),
                                             Filter('id', 'in', [r.target_ref for r in related])])
-        elif obj_type == 'software':
-            sr = self.get_software(match_pattern)
+        elif obj.type == 'software' or obj.type == 'tool':
             software_uses = self.source_handle.query([
                 Filter('type', '=', 'relationship'),
                 Filter('relationship_type', '=', 'uses'),
-                Filter('source_ref', '=', sr.id)
+                Filter('source_ref', '=', obj.id)
             ])
 
             # get the techniques themselves from the ids
             out = self.source_handle.query([
                 Filter('type', '=', 'attack-pattern'),
                 Filter('id', 'in', [r.target_ref for r in software_uses])
             ])
-        elif obj_type == "mitigation":
-            mr = self.get_mitigation(match_pattern)
-            relations = self.source_handle.relationships(mr.id, 'mitigates', source_only=True)
+        elif obj.type == "mitigation":
+            relations = self.source_handle.relationships(obj.id, 'mitigates', source_only=True)
             out = self.source_handle.query([
                 Filter('type', '=', 'attack-pattern'),
                 Filter('id', 'in', [r.target_ref for r in relations])
             ])
-        return self.remove_revoked(out)
+        return remove_revoked(out), obj
```

Comment:
I'm a bit confused by this code.
- Mitigation and group are almost exactly the same. The relationship type param for `source_handle.relationships(` is the only difference as far as I can see, so the code could be consolidated very easily.
- Why is software_uses query not just the same relationships call used by `group` and `mitigation` branches? If it's the same, all three branches could be combined with just a different arg for the relationship type.

Reply:
It took a bit of tweaking, but I managed to get this all condensed down.